Documentation: add migration document (#23751)
cilium-team/cilium: Documentation/installation/k8s-install-migration.rst

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _cni_migration:

*************************************
Migrating a cluster to Cilium
*************************************

Cilium supports "adopting" existing clusters. Running clusters can
be migrated on a node-by-node basis, without disrupting existing traffic

> or requiring a complete cluster outage or rebuild.

Reviewer: I
```suggestion
Cilium can be used to migrate from another cni. Running clusters can
be migrated on a node-by-node basis, without disrupting existing traffic
or requiring a complete cluster outage or rebuild depending on the complexity of the migration case.
```

This document outlines how migrations with Cilium work. You will have a good
understanding of the basic requirements, as well as see an example migration

> which you can practice using :ref:`Kind <gs_kind>`

Reviewer 2: ```suggestion
which you can practice using :ref:`Kind <gs_kind>`.
```

Background
==========

> When a Pod is created, the installed CNI is called which will attach a network
> interface to the Pod. If all goes well, this network interface will join that
> Pod to the cluster network. The network interface shares the same life cycle
> as the Pod, meaning that only newly-created Pods will be picked up by the newer
> CNI if it was swapped out. This requires that all Pods on the cluster that are
> a part of the cluster network be recycled in order to be a member of the new CNI.
> 
> A naive approach to migrating a CNI would be to gradually roll each node on the
> cluster, and then replacing the CNI installed when the node is brought back up.
> This is however, not possible—the second CNI will be installed with a separate
> CIDR range to that of the currently installed CNI, which has no knowledge of
> the other network range. Not to mention the use of different encapsulation
> protocols, or broken network policy since identity is lost.
> 
> Instead, Cilium supports a "hybrid" mode, where two separate overlays are installed
> across the cluster. While pods on a given node can only be attached to one network,
> they have access to both Cilium and non-Cilium pods while the migration is
> taking place. As long as Cilium and the existing networking provider use a separate
> IP range, the Linux routing table takes care of separating traffic.

Reviewer: ```suggestion
When the kubelet creates a Pod, the installed CNI, as configured in /etc/cni/net.d/, is called. The cni will handle the networking for a pod. From allocating an ip address to ensuring that the pod interface is configured appriopriately as part of the pod network. The Pods network configuration shares the same life cycle
as the Pod, In the case of migration, we typically reconfigure /etc/cni/net.d/ to point to the new cni. Once completed any existing pods will have been configured by the old cni and any new pods will be configured by the newer
CNI. To complete the migration all Pods on the cluster that are
configured by the old cni must be recycled in order to be a member of the new CNI.

A naive approach to migrating a CNI would be to reconfigure all nodes with a new CNI and then gradually restart each node in the
cluster, thus replacing the CNI when the node is brought back up and ensuring that all pods are part of the new CNI.

This simple migration while effective comes at the cost of restarting all pods and nodes cluster wide before proceeding. This will be a significant outage for the cluster.

In this document we will discuss a model for live migrating between two deployed CNI implementations. This will have the benefit of reducing downtime of nodes and workloads and ensuring that workloads on both configured CNIs can communicate during migration. 

For live migration to work the second CNI will be installed with a separate
CIDR range to that of the currently installed CNI, which has no knowledge of
the other network range. Not to mention the use of different encapsulation
protocols. If the cluster is making use of network policies or other capabilities that are cni specific. This usually adds significant challenges to the migration model.

To help address the case of live migration as of 1.13 Cilium supports a "hybrid" mode.

This migration strategy allows us to configure each nodes CNI implementation independently. Allowing you to migrate nodes at your own pace and ensuring that workloads on those nodes are able to reach other workloads still deployed using the old CNI while the migration is taking place. 
 As long as Cilium and the existing CNI use a separate
IP range, the Linux routing table takes care of separating traffic.
```

Requirements
============

Live migration requires the following:

- A new, distinct Cluster CIDR for Cilium to use
- Use of the :ref:`Cluster Pool IPAM mode<ipam_crd_cluster_pool>`
- A distinct overlay, either protocol or port
- An existing network plugin that uses the Linux routing stack, such as Flannel, Calico, or AWS-CNI

Limitations
===========

Currently, Cilium migration has not been tested with:

- BGP-based routing
- Changing IP families (e.g. from IPv4 to IPv6)
- Migrating from Cilium in chained mode

During migration, NetworkPolicy and CiliumNetworkPolicy enforcement must
be disabled. Otherwise, traffic from non-Cilium pods may be incorrectly
dropped. Once the migration process is complete, policy enforcement can
be re-enabled.

It is strongly recommended to install Cilium using the :ref:`cluster-pool <ipam_crd_cluster_pool>`

> IPAM allocator.  This provides the strongest assurance that there will

Reviewer 2: ```suggestion
IPAM allocator. This provides the strongest assurance that there will
```

be no IP collisions.

.. warning::
  Migration is highly dependent on the exact configuration of existing
  clusters. It is, thus, strongly recommended to perform a trial migration
  on a test or lab cluster.

Overview
========

The migration process utilizes the :ref:`per-node configuration<per-node-configuration>`
feature to selectively enable Cilium CNI. This allows for a controlled rollout
of Cilium without disrupting existing workloads.

Cilium will be installed, first, in a mode where it establishes an overlay
but does not provide CNI networking for any pods. Then, individual nodes will
be migrated.

In summary, the process looks like:

> 1. Install cilium in "secondary" mode.

Reviewer 2: ```suggestion
1. Install cilium in "secondary" mode
```

2. Cordon, drain, migrate, and reboot each node
3. Remove the existing network provider
4. (Optional) Reboot each node again

Migration procedure
===================

Preparation
-----------

> Optional: Create a :ref:`Kind <gs_kind>` cluster and install Flannel on it

Reviewer 2: ```suggestion
Optional: Create a :ref:`Kind <gs_kind>` cluster and install Flannel on it.
```

  .. code-block:: shell-session

    $ cat <<EOF > kind-config.yaml
    apiVersion: kind.x-k8s.io/v1alpha4
    kind: Cluster
    nodes:
    - role: control-plane
    - role: worker
    - role: worker
    networking:
      disableDefaultCNI: true
    EOF
    $ kind create cluster --config=kind-config.yaml

>     $ kubectl apply -n kube-system --server-side -f \ |SCM_WEB|\/examples/misc/migration/install-reference-cni-plugins.yaml

Reviewer 2: You need this block to be a `.. parsed-literal::`, or the substitution for `|SCM_WEB|` will not happen.

1.  Select a **new** CIDR for pods. It must be distinct from all other CIDRs in use.

    For Kind clusters, the default is ``10.244.0.0/16``. So, for this example, we will
    use ``10.245.0.0/16``.

2.  Select a **distinct** encapsulation port. For example, if the existing cluster
    is using VXLAN, then you should either use GENEVE or configure Cilium to use VXLAN
    with a different port.

    For this example, we will use VXLAN with a non-default port of 8473. But the Cilium
    default of GENEVE encapsulation is preferable.

3.  Create a helm ``values-migration.yaml`` file based on the following example. Be sure to fill
    in the CIDR you selected in step 1.

    .. code-block:: yaml

        operator:
          unmanagedPodWatcher:
            restart: false # Migration: Don't restart unmigrated pods
        tunnel: vxlan # Migration: Optional: default is geneve, configure as needed
        tunnelPort: 8473 # Migration: Optional, change only if both networks use the same port by default
        cni:
          customConf: true # Migration: don't install a CNI configuration file
        ipam:
          mode: "cluster-pool"
          operator:

>             clusterPoolIPv4PodCIDRList: ["10.245.0.0/16"] # Migration: ensure this is distinct and unused

Reviewer 2: ```suggestion
        tunnel: vxlan # Migration: Optional: default is GENEVE, configure as needed
        tunnelPort: 8473 # Migration: Optional, change only if both networks use the same port by default
        cni:
          customConf: true # Migration: Don't install a CNI configuration file
        ipam:
          mode: "cluster-pool"
          operator:
            clusterPoolIPv4PodCIDRList: ["10.245.0.0/16"] # Migration: Ensure this is distinct and unused
```

        policyEnforcementMode: "never" # Migration: Disable policy enforcement
        bpf:
          hostLegacyRouting: true # Migration: Allow for routing between Cilium and the existing overlay

> 4.  Optional: auto-detect settings using the :ref:`cilium-cli <install_cilium_cli>`.

Reviewer 3: I don't think this should be "Optional" as if a reader decided to skip through the steps, they'd miss this step and the creation of the values-initial.yaml file which is required for the following step.

Author: fair enough; updated

    .. code-block:: shell-session

      $ cilium status --wait
      ...
      Cluster Pods:     0/3 managed by Cilium

6.  Create a :ref:`per-node config<per-node-configuration>` that will instruct Cilium to "take over" CNI networking
    on the node. Initially, this will apply to no nodes; you will roll it out gradually via
    the migration process.

    .. code-block:: shell-session

        cat <<EOF | kubectl apply --server-side -f -
        apiVersion: cilium.io/v2alpha1
        kind: CiliumNodeConfig
        metadata:
          namespace: kube-system
          name: cilium-default
        spec:
          nodeSelector:
            matchLabels:
              io.cilium.migration/cilium-default: "true"
          defaults:
            write-cni-conf-when-ready: /host/etc/cni/net.d/05-cilium.conflist
            custom-cni-conf: "false"
            cni-chaining-mode: "none"
            cni-exclusive: "true"
        EOF

Migration
---------

At this point, you are ready to begin the migration process. The basic flow is:

Select a node to be migrated. It is not recommended to start with a control-plane node.

.. code-block:: shell-session

>   export NODE="kind-worker" # for the kind example

Reviewer 2: No need to `export`, I think it's not necessary here (plus the `export` syntax varies between shells).

```suggestion
  $ NODE="kind-worker" # for the Kind example
```

1.  **Optional: cordon and drain the node in question.** Depending on your individual requirements,
    you may wish to cordon and drain before applying the migration.

    **Pros:**

    - Users are isolated from any potential issues
    - Affected pods are moved to a new node and immediately restarted, rather than waiting for the node to reboot

    **Cons:**

    - Pods may be restarted twice during the migration

>     - Any potential issues may be noticed sooner

Reviewer 2: Is this one a disadvantage?

Author: The wording is funny, you're right. I meant to say that potential issues are more likely to show up if you don't drain. I'll re-phrase.

    It is recommended to always cordon and drain at the beginning of the migration process, so that
    end-users are not impacted by any potential issues.

    .. code-block:: shell-session

      $ kubectl cordon $NODE

>       $ kubectl drain $NODE

Reviewer 3: When going through the tutorial, when I drain, I get some error messages draining the node from the Flannel and Cilium Daemonset.

Should we be draining the Node using `--ignore-daemonsets` ?

```
There are pending nodes to be drained:
 kind-worker
cannot delete DaemonSet-managed Pods (use --ignore-daemonsets to ignore): kube-flannel/kube-flannel-ds-rf9b4, kube-system/cilium-2j2lb, kube-system/install-cni-plugins-md4gh, kube-system/kube-proxy-zhxcg
```

Author: Yup, good catch.

> 2.  Label the node. This causes the CiliumNodeConfig to apply to this node.

Reviewer 2: ```suggestion
2.  Label the node. This causes the ``CiliumNodeConfig`` to apply to this node.
```

    .. code-block:: shell-session

      $ kubectl label node $NODE --overwrite "io.cilium.migration/cilium-default=true"

3.  Reboot the node.

    If using kind, simulate a reboot by restarting Cilium on the node:

    .. code-block:: shell-session

>       kubectl -n kube-system delete pod --field-selector spec.nodeName=$NODE -l k8s-app=cilium

Reviewer 2: ```suggestion
    If using Kind, simulate a reboot by restarting Cilium on the node:

    .. code-block:: shell-session

      $ kubectl -n kube-system delete pod --field-selector spec.nodeName=$NODE -l k8s-app=cilium
```

Reviewer: you can simulate a node reboot with docker restart `kind-node`

Reviewer: and you should cause this clears out old not in use interfaces and erroneuos iptables things

.. only:: not (epub or latex or html)

    WARNING: You are looking at unreleased Cilium documentation.
    Please use the official rendered version released here:
    https://docs.cilium.io

.. _cni_migration:

*************************************
Migrating a cluster to Cilium
*************************************

Cilium can be used to migrate from another cni. Running clusters can
be migrated on a node-by-node basis, without disrupting existing traffic
or requiring a complete cluster outage or rebuild depending on the complexity of the migration case.

This document outlines how migrations with Cilium work. You will have a good
understanding of the basic requirements, as well as see an example migration
which you can practice using :ref:`Kind <gs_kind>`.

Background
==========

When the kubelet creates a Pod's Sandbox, the installed CNI, as configured in ``/etc/cni/net.d/``,
is called. The cni will handle the networking for a pod - including allocating 
an ip address, creating & configuring a network interface, and (potentially)

> establishing an overlay network. The Pod'ss network configuration shares the

Reviewer 2: Oops, the spell checker found a typo

```suggestion
establishing an overlay network. The Pod's network configuration shares the
```

Author: oh no, i totally meant that :-p

      $ kubectl apply --server-side -f https://github.com/flannel-io/flannel/releases/latest/download/kube-flannel.yml
      $ kubectl wait --for=condition=Ready nodes --all

- Optional: Monitor connectivity.

  You may wish to install a tool such as `goldpinger <https://github.com/bloomberg/goldpinger>`_
  to detect any possible connectivity issues.

1.  Select a **new** CIDR for pods. It must be distinct from all other CIDRs in use.

    For Kind clusters, the default is ``10.244.0.0/16``. So, for this example, we will
    use ``10.245.0.0/16``.

2.  Select a **distinct** encapsulation port. For example, if the existing cluster
    is using VXLAN, then you should either use GENEVE or configure Cilium to use VXLAN
    with a different port.

    For this example, we will use VXLAN with a non-default port of 8473. But the Cilium
    default of GENEVE encapsulation is preferable.

3.  Create a helm ``values-migration.yaml`` file based on the following example. Be sure to fill
    in the CIDR you selected in step 1.

    .. code-block:: yaml

        operator:
          unmanagedPodWatcher:
            restart: false # Migration: Don't restart unmigrated pods

>         tunnel: vxlan # Migration: Optional: default is GENEVE, configure as needed

Reviewer 3: Is Geneve definitely the default? Helm default value is VXLAN:
https://github.com/cilium/cilium/blob/c9723a8df3cfa336da1f8457a864105d8349acfe/install/kubernetes/cilium/values.yaml#L1681

Author: Oh, you're totally right. Whoops.

    a brief interruption while the node is rebooted.

2.  Label the node. This causes the ``CiliumNodeConfig`` to apply to this node.

    .. code-block:: shell-session

      $ kubectl label node $NODE --overwrite "io.cilium.migration/cilium-default=true"

3.  Restart Cilium. This will cause it to write its CNI configuration file.

    .. code-block:: shell-session

      $ kubectl -n kube-system delete pod --field-selector spec.nodeName=$NODE -l k8s-app=cilium
      $ kubectl -n kube-system rollout status ds/cilium -w

4.  Reboot the node.

    If using kind, do so with docker:

    .. code-block:: shell-session
    
      docker restart $NODE

5.  Validate that the node has been successfully migrated.

    .. code-block:: shell-session

      $ cilium status --wait
      $ kubectl get -o wide node $NODE
      $ kubectl -n kube-system run --attach --rm --restart=Never verify-network \
        --overrides='{"spec": {"nodeName": "'$NODE'", "tolerations": [{"operator": "Exists"}]}}' \
        --image ghcr.io/nicolaka/netshoot:v0.8 -- /bin/bash -c 'ip -br addr && curl -s -k https://$KUBERNETES_SERVICE_HOST/healthz && echo'

>     Ensure the IP address of the pod is in the Cilium CIDR(s) supplied above.

Reviewer: Ensure the IP address of the pod is in the Cilium CIDR(s) supplied above and that the apiservice is reachable.
